ttnn/cpp/ttnn/operations/conv2d/device/kernels/activation_reader_width_sharded.cpp

```diff
@@ -0,0 +1,232 @@
+
+        //After reading one block, increment the starting read pointer by the width of the block.
+        //Next read uses the next set of channels.
+        act_l1_read_addr +=conv_act_c_read_bytes;
+
+        noc_async_read_barrier();
+        cb_push_back(cb_id_act_row_major_bfloat16, act_block_num_tiles);
+
+
+        // Round robin self-mcast and receive tilized act matrix in cb_id_act
+        // Compute should function like regular mm
+
+        uint32_t act_w_outer_i = 0;
+        for (uint32_t act_w_outer_i = 0; act_w_outer_i < act_w_num_outer; act_w_outer_i++) {
+            cb_reserve_back(cb_id_act, act_block_num_tiles);
+            if (act_w_outer_i == act_mcast_sender_id) {
+                // MCAST SENDER: send entire tilized input to other cores in column
+                // wait until all act mcast destinations have atomically incremented the act semaphore_addr (i.e. its value should be act_mcast_num_dests), then reset
+                // the semaphore_addr value back to zero for the next block
+
+                noc_semaphore_wait_min(act_mcast_sender_semaphore_addr_ptr, act_mcast_num_cores-1);
+                noc_semaphore_set(act_mcast_sender_semaphore_addr_ptr, 0);
+
+                noc_semaphore_set(act_mcast_receiver_semaphore_addr_ptr, INVALID);
+
+                // // compute tilizes and pops cb_id_act and pushes to tilized_in0_cb_id
+                cb_wait_front(tilized_in0_cb_id, act_block_num_tiles);
+
+                // // Now we have the block in the CB address, we can mcast to dests!
+                uint32_t tilized_act_start_address = get_read_ptr(tilized_in0_cb_id);
+
+                uint64_t act_multicast_data_addr = act_multicast_noc_addr | get_write_ptr(cb_id_act);
+                // // num_dests will source, since we are copying to a different local CB as well
+                noc_async_write_multicast_loopback_src(tilized_act_start_address, act_multicast_data_addr, act_mcast_sender_size_bytes, 16 , false, false);
+
+
+                // Note: no need for write barrier, since these two multicasts are done on the same noc id, same vc, same cmd_buf
+                // Also, this only works because we are setting VCs statically (using NOC_CMD_STATIC_VC).
+
+                // // We should also multicast VALID flag to destinations for receiver semaphore
+                noc_semaphore_set_multicast_loopback_src(act_mcast_sender_semaphore_valid_addr, act_mcast_receiver_semaphore_noc_addr, 16 , false, false);
+
+
+                noc_semaphore_wait(act_mcast_receiver_semaphore_addr_ptr, VALID);
+
+            } else {
+
+                // MCAST RECEIVER: receive entire tilized input from sender core
+                // Set act semaphore value to INVALID
+                noc_semaphore_set(act_mcast_receiver_semaphore_addr_ptr, INVALID);
+
+                uint32_t sender_x = act_mcast_x_lookup[act_w_outer_i%num_cores_x];
+                uint32_t sender_y = act_mcast_y_lookup[act_w_outer_i/num_cores_x];
```

Comment:
`/` and `%` would be expensivein innermost loop, any way to avoid? using some counters?

Reply:
Fixed.